### PR TITLE
Add `AdjacentListsSupport` plugin to separate lists of the same type

### LINKED ISSUE #12738: New list plugin doesn't convert subsequent lists with a different type properly (aka resetting numbering, splitting lists)
## 📝 Provide detailed reproduction steps (if any)

For context this is a simplified markup produced by MS Word.

1.  Copy following HTML to your clipboard:

```html
<p style="margin-top: 0px; margin-bottom: 0px; line-height: 1">foo</p>

<p style="margin-top: 0px; margin-bottom: 0px; margin-left: 24px; line-height: 1">&nbsp;</p>
<ol style="list-style-type: decimal">
    <li>

        <p style="margin-top: 0px; margin-bottom: 10.67px; line-height: 1.05">foobar</p>
    </li>
</ol>

<p style="margin-top: 0px; margin-bottom: 10.67px; margin-left: 24px; line-height: 1.05">&nbsp;</p>
<ol style="list-style-type: lower-latin">
    <li>

        <p style="margin-top: 0px; margin-bottom: 10.67px; line-height: 1.05">foobar</p>
    </li>
    <li>

        <p style="margin-top: 0px; margin-bottom: 10.67px; line-height: 1.05">foobar</p>
    </li>
</ol>
<ol start="2" style="list-style-type: decimal">
    <li>

        <p style="margin-top: 0px; margin-bottom: 10.67px; line-height: 1.05">Decimal list item again</p>
    </li>
    <li>

        <p style="margin-top: 0px; margin-bottom: 10.67px; line-height: 1.05">Decimal list item again
        </p>
    </li>
</ol>

<p style="margin-top: 0px; margin-bottom: 10.67px; margin-left: 48px; line-height: 1.05">&nbsp;</p>
<ol style="list-style-type: lower-latin">
    <li>

        <p style="margin-top: 0px; margin-bottom: 10.67px; line-height: 1.05">foobar</p>
    </li>
    <li>

        <p style="margin-top: 0px; margin-bottom: 10.67px; line-height: 1.05">foobar</p>
    </li>
</ol>
<ul style="list-style-type: disc">
    <li>

        <p style="margin-top: 0px; margin-bottom: 10.67px; line-height: 1.05">foobar</p>
    </li>
</ul>
<ol start="3" style="list-style-type: lower-latin">
    <li>

        <p style="margin-top: 0px; margin-bottom: 10.67px; line-height: 1.05">foobar
        </p>
    </li>
</ol>
```

1.  Open a demo that has new list (a.k.a. document list) implementation in it. [Import from Word guide](https://ckeditor.com/docs/ckeditor5/latest/features/import-word/import-word.html) is a good example.
2.  Open js dev console.
3.  Execute ``editor.setData( `<copiedMarkupHere>` )``.

### ✔️ Expected result

List with two "Decimal list item again" items should be a numbered list.

### ❌ Actual result

List goes on as alphanum list.

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by filipsobol**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Feature(list): Add `AdjacentListsSupport` plugin to separate lists of the same type. Closes #12738.

### CODE REVIEW

File: packages/ckeditor5-list/src/documentlist/documentlistseparator.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+		const editor = this.editor;
+		const model = editor.model;
+
+		model.schema.register( 'listSeparator', {
+			allowWhere: '$block',
+			isBlock: true
+		} );
+
+		editor.conversion.for( 'upcast' )
+			// Add `listSeparator` element between similar list elements on upcast
```
Comment (niegowski) [inline packages/ckeditor5-list/src/documentlist/documentlistseparator.ts:35]:
```suggestion
			// Add `listSeparator` element between similar list elements on upcast.
```

File: packages/ckeditor5-list/src/documentlist/documentlistseparator.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
+/**
+ * @license Copyright (c) 2003-2023, CKSource Holding sp. z o.o. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module list/documentlist/documentlistutils
+ */
```
Comment (niegowski) [inline packages/ckeditor5-list/src/documentlist/documentlistseparator.ts:8]:
```suggestion
 */
 
```

File: packages/ckeditor5-list/src/documentlist/documentlistseparator.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
+/**
+ * @license Copyright (c) 2003-2023, CKSource Holding sp. z o.o. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module list/documentlist/documentlistutils
```
Comment (niegowski) [inline packages/ckeditor5-list/src/documentlist/documentlistseparator.ts:7]:
```suggestion
 * @module list/documentlist/documentlistseparator
```

File: packages/ckeditor5-list/src/documentlist/documentlistseparator.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+ */
+import type { GetCallback } from 'ckeditor5/src/utils';
+import { Plugin } from 'ckeditor5/src/core';
+
+import type {	UpcastElementEvent, ViewElement } from 'ckeditor5/src/engine';
```
Comment (niegowski) [inline packages/ckeditor5-list/src/documentlist/documentlistseparator.ts:12]:
```suggestion
import type { UpcastElementEvent, ViewElement } from 'ckeditor5/src/engine';
```

File: packages/ckeditor5-list/src/documentlist/documentlistseparator.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+			isBlock: true
+		} );
+
+		editor.conversion.for( 'upcast' )
+			// Add `listSeparator` element between similar list elements on upcast
+			.add( dispatcher => {
+				dispatcher.on<UpcastElementEvent>( 'element:ol', listSeparatorUpcastConverter() );
+				dispatcher.on<UpcastElementEvent>( 'element:ul', listSeparatorUpcastConverter() );
+			} )
+			// View to model transformation
```
Comment (niegowski) [inline packages/ckeditor5-list/src/documentlist/documentlistseparator.ts:40]:
```suggestion
			// View to model transformation.
```

File: packages/ckeditor5-list/src/documentlist/documentlistseparator.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+			.elementToElement( {
+				model: 'listSeparator',
+				view: 'ck-list-separator'
+			} );
+
+		// `listSeparator` should exist in view, but be invisible (hidden)
```
Comment (niegowski) [inline packages/ckeditor5-list/src/documentlist/documentlistseparator.ts:46]:
```suggestion
		// The `listSeparator` element should exist in the view, but should be invisible (hidden).
```

File: packages/ckeditor5-list/src/documentlist/documentlistseparator.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+		editor.conversion.for( 'editingDowncast' ).elementToElement( {
+			model: 'listSeparator',
+			view: {
+				name: 'div',
+				classes: [ 'ck-list-separator', 'ck-hidden' ]
+			}
+		} );
+
+		// `listSeparator` should not exist in output data
```
Comment (niegowski) [inline packages/ckeditor5-list/src/documentlist/documentlistseparator.ts:55]:
```suggestion
		// The `listSeparator` element should not exist in output data.
```

File: packages/ckeditor5-list/tests/documentlist/documentlistseparator.js [hunk 8]
```diff
@@ -0,0 +1,300 @@
+/**
+ * @license Copyright (c) 2003-2023, CKSource Holding sp. z o.o. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/* globals document */
```
Comment (niegowski) [inline packages/ckeditor5-list/tests/documentlist/documentlistseparator.js:6]:
```suggestion
/* globals document */

```

File: packages/ckeditor5-list/tests/documentlist/documentlistseparator.js [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+
+	describe( 'upcast', () => {
+		it( 'inserts "listSeparator" element between two "ul" lists in model', () => {
+			editor.setData(
+				'<ul>' +
+				'<li>One</li>' +
+				'</ul>' +
+				'<ul>' +
+				'<li>Two</li>' +
+				'</ul>'
```
Comment (niegowski) [inline packages/ckeditor5-list/tests/documentlist/documentlistseparator.js:70]: Please fix the code formatting here and in other tests.

**Review by niegowski — CHANGES_REQUESTED** [review]

**Comment by Reinmar** [comment]
The proposed plugin name (`DocumentListSeparator`) leaks the implementation detail of the plugin (the use of a separator).

For a consumer, this feature is about having adjacent lists of the same type or resetting numbering. Maybe we could look for something more integrator-oriented?

File: packages/ckeditor5-list/tests/documentlist/documentlistseparator.js (hunk 9 above)
Comment (filipsobol) [reply packages/ckeditor5-list/tests/documentlist/documentlistseparator.js]: Done :heavy_check_mark:

**Comment by arkflpc** [comment]
> The proposed plugin name (`DocumentListSeparator`) leaks the implementation detail of the plugin (the use of a separator).
> 
> For a consumer, this feature is about having adjacent lists of the same type or resetting numbering. Maybe we could look for something more integrator-oriented?

Idea: `AdjacentListSupport`.

File: packages/ckeditor5-list/docs/features/document-lists.md [hunk 10]
```diff
@@ -96,6 +96,28 @@ ClassicEditor
...
+## List merging
+
+By default, two lists of the same type (ordered and unordered) that are next to each other are merged together. This is done so that lists that visually appear to be one continuous list actually are, even if the user has accidentally created several of them.
+
+Unfortunately, in some cases this can be undesirable behavior. For example, two adjacent numbered lists, each with two items, will merge into a single list with the numbers 1 through 4.
+
+To prevent this behavior, enable the `AdjacentListsSupport` plugin.
+
+```js
+import AdjacentListsSupport from '@ckeditor/ckeditor5-list/src/documentlist/adjacentlistssupport.js';
```
Comment (niegowski) [inline packages/ckeditor5-list/docs/features/document-lists.md:108]: This should import from index file.

**Review by niegowski — APPROVED** [review]

**Pull request merged**